File: VB/GridLookUpEditCBMultipleSelection/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Repository

Namespace GridLookUpEditCBMultipleSelection
	Partial Public Class Form1
		Inherits Form
		Private gridCheckMarksSA As GridCheckMarksSelection
		Private gridCheckMarksIP As GridCheckMarksSelection

		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			Dim dt As DataTable = FillDataTable()
			gridLookUpEdit1.Properties.DataSource = dt
			gridLookUpEdit1.Properties.DisplayMember = "Fruit"
			gridLookUpEdit1.Properties.View.OptionsSelection.MultiSelect = True
			AddHandler gridLookUpEdit1.CustomDisplayText, AddressOf gridLookUpEdit1_CustomDisplayText
			gridLookUpEdit1.Properties.PopulateViewColumns()

			repositoryItemGridLookUpEdit1.DataSource = dt
			repositoryItemGridLookUpEdit1.DisplayMember = "Fruit"
			repositoryItemGridLookUpEdit1.View.OptionsSelection.MultiSelect = True
			AddHandler repositoryItemGridLookUpEdit1.CustomDisplayText, AddressOf gridLookUpEdit1_CustomDisplayText
			repositoryItemGridLookUpEdit1.PopulateViewColumns()



			gridCheckMarksSA = New GridCheckMarksSelection(gridLookUpEdit1.Properties)
			AddHandler gridCheckMarksSA.SelectionChanged, AddressOf gridCheckMarks_SelectionChanged
			gridCheckMarksSA.SelectAll(dt.DefaultView)
			gridLookUpEdit1.Properties.Tag = gridCheckMarksSA

			gridCheckMarksIP = New GridCheckMarksSelection(repositoryItemGridLookUpEdit1)
			repositoryItemGridLookUpEdit1.Tag = gridCheckMarksIP
			AddHandler gridCheckMarksIP.SelectionChanged, AddressOf gridCheckMarks_SelectionChanged
		End Sub

		Private Sub gridCheckMarks_SelectionChanged(ByVal sender As Object, ByVal e As EventArgs)
			If TypeOf ActiveControl Is GridLookUpEdit Then
				Dim sb As New StringBuilder()
				For Each rv As DataRowView In (TryCast(sender, GridCheckMarksSelection)).Selection
					If sb.ToString().Length > 0 Then
						sb.Append(", ")
					End If
					sb.Append(rv("Fruit").ToString())
				Next rv
				TryCast(ActiveControl, GridLookUpEdit).Text = sb.ToString()
			End If
		End Sub

		Private Sub gridLookUpEdit1_CustomDisplayText(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.Controls.CustomDisplayTextEventArgs)
			Dim sb As New StringBuilder()
			Dim gridCheckMark As GridCheckMarksSelection = If(TypeOf sender Is GridLookUpEdit, TryCast((TryCast(sender, GridLookUpEdit)).Properties.Tag, GridCheckMarksSelection), TryCast((TryCast(sender, RepositoryItemGridLookUpEdit)).Tag, GridCheckMarksSelection))
			If gridCheckMark Is Nothing Then
				Return
			End If
			For Each rv As DataRowView In gridCheckMark.Selection
				If sb.ToString().Length > 0 Then
					sb.Append(", ")
				End If
				sb.Append(rv("Fruit").ToString())
			Next rv
			e.DisplayText = sb.ToString()
		End Sub

		Private Function FillDataTable() As DataTable
			Dim _dataTable As New DataTable()
			Dim col As DataColumn
			Dim row As DataRow

			col = New DataColumn()
			col.ColumnName = "Fruit"
			col.DataType = System.Type.GetType("System.String")
			_dataTable.Columns.Add(col)

			row = _dataTable.NewRow()
			row("Fruit") = "Peach"
			_dataTable.Rows.Add(row)
			row = _dataTable.NewRow()
			row("Fruit") = "Apple"
			_dataTable.Rows.Add(row)
			row = _dataTable.NewRow()
			row("Fruit") = "Banana"
			_dataTable.Rows.Add(row)

			Return _dataTable
		End Function

		Private Sub barEditItem1_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs) Handles barEditItem1.EditValueChanged

		End Sub
	End Class
End Namespace

File: CS/GridLookUpEditCBMultipleSelection/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Repository;

namespace GridLookUpEditCBMultipleSelection
{
    public partial class Form1 : Form
    {
        GridCheckMarksSelection gridCheckMarksSA;
        GridCheckMarksSelection gridCheckMarksIP;

        public Form1()
        { InitializeComponent(); }

        private void Form1_Load(object sender, EventArgs e)
        {
            DataTable dt = FillDataTable();
            gridLookUpEdit1.Properties.DataSource = dt;
            gridLookUpEdit1.Properties.DisplayMember = "Fruit";
            gridLookUpEdit1.Properties.View.OptionsSelection.MultiSelect = true;
            gridLookUpEdit1.CustomDisplayText += new DevExpress.XtraEditors.Controls.CustomDisplayTextEventHandler(gridLookUpEdit1_CustomDisplayText);
            gridLookUpEdit1.Properties.PopulateViewColumns();

            repositoryItemGridLookUpEdit1.DataSource = dt;
            repositoryItemGridLookUpEdit1.DisplayMember = "Fruit";
            repositoryItemGridLookUpEdit1.View.OptionsSelection.MultiSelect = true;
            repositoryItemGridLookUpEdit1.CustomDisplayText += new DevExpress.XtraEditors.Controls.CustomDisplayTextEventHandler(gridLookUpEdit1_CustomDisplayText);
            repositoryItemGridLookUpEdit1.PopulateViewColumns();
            


            gridCheckMarksSA = new GridCheckMarksSelection(gridLookUpEdit1.Properties);
            gridCheckMarksSA.SelectionChanged += new GridCheckMarksSelection.SelectionChangedEventHandler(gridCheckMarks_SelectionChanged);
            gridCheckMarksSA.SelectAll(dt.DefaultView);
            gridLookUpEdit1.Properties.Tag = gridCheckMarksSA;

            gridCheckMarksIP = new GridCheckMarksSelection(repositoryItemGridLookUpEdit1);
            repositoryItemGridLookUpEdit1.Tag = gridCheckMarksIP;
            gridCheckMarksIP.SelectionChanged += new GridCheckMarksSelection.SelectionChangedEventHandler(gridCheckMarks_SelectionChanged);
        }

        void gridCheckMarks_SelectionChanged(object sender, EventArgs e)
        {
            if (ActiveControl is GridLookUpEdit)
            {
                StringBuilder sb = new StringBuilder();
                foreach (DataRowView rv in (sender as GridCheckMarksSelection).Selection)
                {
                    if (sb.ToString().Length > 0) { sb.Append(", "); }
                    sb.Append(rv["Fruit"].ToString());
                }
                (ActiveControl as GridLookUpEdit).Text = sb.ToString();
            }            
        }

        void gridLookUpEdit1_CustomDisplayText(object sender, DevExpress.XtraEditors.Controls.CustomDisplayTextEventArgs e)
        {
            StringBuilder sb = new StringBuilder();
            GridCheckMarksSelection gridCheckMark = sender is GridLookUpEdit ? (sender as GridLookUpEdit).Properties.Tag as GridCheckMarksSelection : (sender as RepositoryItemGridLookUpEdit).Tag as GridCheckMarksSelection;
            if (gridCheckMark == null) return;
            foreach (DataRowView rv in gridCheckMark.Selection)
            {
                if (sb.ToString().Length > 0) { sb.Append(", "); }
                sb.Append(rv["Fruit"].ToString());
            }
            e.DisplayText = sb.ToString();
        }

        DataTable FillDataTable()
        {
            DataTable _dataTable = new DataTable();
            DataColumn col;
            DataRow row;

            col = new DataColumn();
            col.ColumnName = "Fruit";
            col.DataType = System.Type.GetType("System.String");
            _dataTable.Columns.Add(col);

            row = _dataTable.NewRow();
            row["Fruit"] = "Peach";
            _dataTable.Rows.Add(row);
            row = _dataTable.NewRow();
            row["Fruit"] = "Apple";
            _dataTable.Rows.Add(row);
            row = _dataTable.NewRow();
            row["Fruit"] = "Banana";
            _dataTable.Rows.Add(row);

            return _dataTable;
        }

        private void barEditItem1_EditValueChanged(object sender, EventArgs e)
        {

        }
    }
}
